### ISSUE TITLE
Single versus double based numerics

### ISSUE THREAD

**Issue by mwpowellhtx** [martindevans/SharpSteer2#14]
Hello,

The code seems to be heavily Vector3-oriented, which makes sense to represent such things as vectors in three dimensional space. What is the advantage of using a [Single-based numerics vector](http://msdn.microsoft.com/en-us/library/system.numerics.vector3.aspx), as compared/contrasted with Double-based one?

Besides, it seems to be the only choice in the numerics .NET namespace. Is this for precision reasons? Why wouldn't double-based be more efficient? Not least of which from a mapping to data store perspective, as in my case.

Also, once upon a time this might have been the case that Single/float was more efficient than Double/double, but any more it isn't the case. Or have I missed something?

Thank you...

**Comment by martindevans** [comment]
There's a couple of reasons that the library is single based.

Historically the engine has been ported through a variety of versions and all of those versions have been single precision - so the initial design decision was made years ago. I could port it to double precision, so the question is really "what is the advantage of using a double precision vector?".

As you say, the only choice for vectors in System.Net.Numerics is single precision vectors. This is because these vector types are SIMD enabled - on hardware with the right instructions vector operations will be done on all three elements in one clock cycle. However SIMD can only operate on certain size vectors, I'm not aware of any hardware which could operate on 192 bit vectors (3x64 bit)! Single precision may only be marginally faster than double without SIMD, but it win hands down with SIMD.

Finally the major game development platform for C# is unity. Unity uses single precision floats so any more precision within SS2 would be spurious.

tl;dr - What specific problem are you facing where you think double precision would be helpful? I'll do what I can to help.

**Comment by mwpowellhtx** [comment]
Hi, thanks for the reference on the SIMD. Yes, I noticed that, but didn't quite understand. But this helps clarify, thank you.

None with the Steer API just yet. Up to now, for quantities, dimensional calculations, etc, I've been assuming Double precision, but it would be doable to support single as well, probably.

These are definitely good reasons to want better performing floating point calculations.
